cvmfs/sqlitemem.cc

```diff
@@ -0,0 +1,489 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#define __STDC_FORMAT_MACROS
+
+#include "cvmfs_config.h"
+#include "sqlitemem.h"
+
+#include <cassert>
+#include <cstddef>
+#include <cstring>
+#include <new>
+
+#include "smalloc.h"
+#include "util_concurrency.h"
+
+using namespace std;  // NOLINT
+
+namespace sqlite {
+
+uint32_t MemoryManager::MallocArena::GetSize(void *ptr) const {
+  assert(Contains(ptr));
+
+  ReservedBlockCtl *block_ctl = reinterpret_cast<ReservedBlockCtl *>(
+    reinterpret_cast<char *>(ptr) - sizeof(ReservedBlockCtl));
```

Comment:
Perhaps factor that out into an inlinable method like `GetReservedBlockCtl(void *ptr)`?